src/networking.c

```diff
@@ -4317,44 +4318,70 @@ void killIOThreads(void) {
-        pthread_mutex_lock(&io_threads_mutex[j]);
-    server.io_threads_active = 0;
+    for (int i = 0; i < num; i++)
+        pthread_mutex_lock(&io_threads_mutex[server.io_threads_active_num -1 - i]);
+    server.io_threads_active_num -= num;
 }
 
-/* This function checks if there are not enough pending clients to justify
- * taking the I/O threads active: in that case I/O threads are stopped if
- * currently active. We track the pending writes as a measure of clients
- * we need to handle in parallel, however the I/O threading is disabled
- * globally for reads as well if we have too little pending clients.
+/* This function will dynamically adjust the I/O threads number according to
+ * real workloads in runtime, currently we track the latest x rounds of
+ * pending writes as a measure of clients we need to handle in parallel,
+ * however the I/O threading is disabled globally for reads as well if
+ * we have too little pending clients.
  *
  * The function returns 0 if the I/O threading should be used because there
  * are enough active threads, otherwise 1 is returned and the I/O threads
  * could be possibly stopped (if already active) as a side effect. */
-int stopThreadedIOIfNeeded(void) {
-    int pending = listLength(server.clients_pending_write);
-
-    /* Return ASAP if IO threads are disabled (single threaded mode). */
+#define IO_THREAD_BATCH_NUM 2
+int adjustThreadedIOIfNeeded(void) {
+    /* Return ASAP if I/O threads are disabled (single threaded mode). */
     if (server.io_threads_num == 1) return 1;
 
-    if (pending < (server.io_threads_num*2)) {
-        if (server.io_threads_active) stopThreadedIO();
-        return 1;
+    int pending = listLength(server.clients_pending_write);
+    server.clients_pending_write_avg_num = (int)(server.clients_pending_write_avg_num * 0.8 +
+        pending * 0.2);
+
+    /* I/O threads number except the main thread. */
+    int ioThreadsNow = server.io_threads_active_num - 1;
+    int ioThreadsReq = server.clients_pending_write_avg_num / IO_THREAD_BATCH_NUM;
+    int ioThreadsCap = server.io_threads_num - server.io_threads_active_num;
+
+    if (ioThreadsReq < ioThreadsNow) {
+        /* The requested I/O threads is less than existing active I/O threads,
+         * we need to deactivate some I/O threads to make CPU more efficiency. */
+        if (server.io_threads_active_num > 1) {
+            if (ioThreadsNow > ioThreadsReq) {
+                int stopIOThreadsNum = ioThreadsNow - ioThreadsReq;
+                deactivateThreadedIO(stopIOThreadsNum);
+            }
+            if (server.io_threads_active_num == 1) return 1;
+        }
+    } else if (ioThreadsReq > ioThreadsNow) {
+        /* We need more I/O threads to increase the parallelism. */
+        if (server.io_threads_active_num < server.io_threads_num) {
+            int startIOThreadsNum = min(ioThreadsReq - ioThreadsNow, ioThreadsCap);
+            activateThreadedIO(startIOThreadsNum);
+        }
```

Comment:
I think the readability can be improved a bit here. consider the following?

```suggestion
    if (server.io_threads_active_num < target_num) {
        /* Increasing active threads */
        for (size_t i = server.io_threads_active_num; i < target_num; ++i) {
            pthread_mutex_unlock(&io_threads_mutex[i]);
        }
    } else {
        /* Decreasing active threads */
        for (size_t i = server.io_threads_active_num; i > target_num; --i) {
            pthread_mutex_lock(&io_threads_mutex[i - 1]);
        }
    }
```